### PR TITLE
add check about exec + diagnostics output

### LINKED ISSUE #1341: Lychee using .exec() even if it is not available.
### Description of the problem

Lychee says php.exec() is optional, but it is used in one step of the installation and later used by exif when uploading a photo.

### Situation 1:
Install lychee from the 4.0.0 release with php.exec() function disabled (In my case it is disabled by the system administrator and it cannot be activated) and get an error with 500 status code on the final step of the installation (after configuring the .env variables from the GUI).

I modified the following files to bypass the version check using exec(). These were the files that caused problems in proceding with the installation process mentioned above.

1.-Lychee/database/migrations/2019_12_25_0600_config_exiftool_ternary.php
    - failed on line 23
    
2.- Lychee/database/migrations/2020_01_04_1200_config_has_ffmpeg.php
    - failed on line 24

I knew this because the error page showing the status code 500 mentioned one of these files. I tried to bypass one, then the error page reapeared but this time pointing the error to a different file. It stoped after modifying the second file.

Then i could install lychee with no problems.

### Situation 2:

1.- Try to upload a photo in any directory using the super admin account.

When trying to upload a photo, exif uses exec() to get the path of the tool it is going to use.
https://github.com/LycheeOrg/php-exif/blob/568b56c6b33cca87dcb56103d1d1eb0609091def/lib/PHPExif/Adapter/Exiftool.php
on line 99

### Output of the diagnostics [REQUIRED]
 Diagnostics
    -----------
    Warning: Dropbox import not working. dropbox_key is empty.
    Warning: You may experience problems when uploading a photo of large size. Take a look in the FAQ for details.
    Warning: You may experience problems when uploading a photo of large size. Take a look in the FAQ for details.
    Warning: You may experience problems when uploading a photo of large size or handling many/large albums. Take a look in the FAQ for details.
    Warning: exec function has been disabled. You may experience some error 500, please report them to us.
    Error: PHP zip extension not activated

    System Information
    ------------------
    Lychee Version (release):        4.4.0
    DB Version:                      4.4.0
    
    composer install:                --no-dev
    APP_ENV:                         production
    APP_DEBUG:                       true
    
    System:                          Linux
    PHP Version:                     8.1
    PHP User agent:                  Lychee/4 (https://lycheeorg.github.io/)
    Max uploaded file size:          2M
    Max post size:                   8M
    Max execution time:              30
    MySQL Version:                   5.7.38
    
    Imagick:                         1
    Imagick Active:                  1
    Imagick Version:                 1690
    GD Version:                      bundled (2.1.0 compatible)

    Config Information
    ------------------
    version:                         040400
    check_for_updates:               0
    sorting_Photos_col:              taken_at
    sorting_Photos_order:            ASC
    sorting_Albums_col:              max_taken_at
    sorting_Albums_order:            ASC
    imagick:                         1
    skip_duplicates:                 0
    small_max_width:                 0
    small_max_height:                360
    medium_max_width:                1920
    medium_max_height:               1080
    lang:                            es
    layout:                          1
    image_overlay_type:              desc
    default_license:                 none
    compression_quality:             90
    full_photo:                      1
    delete_imported:                 0
    Mod_Frame:                       1
    Mod_Frame_refresh:               30
    thumb_2x:                        1
    small_2x:                        1
    medium_2x:                       1
    landing_page_enable:             0
    landing_owner:                   John Smith
    landing_title:                   John Smith
    landing_subtitle:                Cats, Dogs & Humans Photography
    landing_facebook:                https://www.facebook.com/JohnSmith
    landing_flickr:                  https://www.flickr.com/JohnSmith
    landing_twitter:                 https://www.twitter.com/JohnSmith
    landing_instagram:               https://instagram.com/JohnSmith
    landing_youtube:                 https://www.youtube.com/JohnSmith
    landing_background:              dist/cat.jpg
    site_title:                      Lychee v4
    site_copyright_enable:           1
    site_copyright_begin:            2019
    site_copyright_end:              2019
    additional_footer_text:          
    display_social_in_gallery:       0
    public_search:                   0
    SL_enable:                       0
    SL_for_admin:                    0
    public_recent:                   0
    recent_age:                      1
    public_starred:                  0
    downloadable:                    0
    photos_wraparound:               1
    map_display:                     0
    zip64:                           1
    map_display_public:              0
    map_provider:                    Wikimedia
    force_32bit_ids:                 0
    map_include_subalbums:           0
    update_check_every_days:         3
    has_exiftool:                    1
    share_button_visible:            0
    import_via_symlink:              0
    has_ffmpeg:                      1
    location_decoding:               0
    location_decoding_timeout:       30
    location_show:                   1
    location_show_public:            0
    rss_enable:                      0
    rss_recent_days:                 7
    rss_max_items:                   100
    prefer_available_xmp_metadata:   0
    editor_enabled:                  1
    lossless_optimization:           0
    swipe_tolerance_x:               150
    swipe_tolerance_y:               250
    local_takestamp_video_formats:   .avi|.mov
    log_max_num_line:                1000
    unlock_password_photos_with_url_param: 0
    nsfw_visible:                    1
    nsfw_blur:                       0
    nsfw_warning:                    0
    nsfw_warning_admin:              0
    map_display_direction:           1
    album_subtitle_type:             oldstyle
    upload_processing_limit:         4
    public_photos_hidden:            1
    new_photos_notification:         0

### Browser and system
Lychee hosted on Centos 7

### LINKED ISSUE #1396: switiching to current version - installation error 500 exiftool [was: No settings entry in main cog wheel menu]
### Detailed description of the problem [REQUIRED]
Beeing a beginner it might not be a bug but rather an installation issue  hard to manage by me#

There is absolutly no setting as logged in user in the main top left menu, only: change login, sorting, dropbox, about, diagnostics, log, sign out.

Cheers to anyone helping out to get me up to speed. I wanted to dismiss the idea of installing my own foto lib but playing around I found lychee really neat, stylish and lightweight.

### Steps to reproduce the issue
**Steps to reproduce the behavior:**
1. Go to main menu
2. no settings entry

**Screenshots** 
not needed

### Output of the diagnostics  [REQUIRED]
Diagnostics
-----------
Warning: Dropbox import not working. No property for dropboxKey.
No critical problems found. Lychee should work without problems!

System Information
------------------
Lychee Version:  3.1.6
DB Version:      030102
System:          Linux
PHP Version:     8
MySQL Version:   100335
Imagick:         1
Imagick Active:  1
Imagick Version: 1690
GD Version:      bundled (2.1.0 compatible)

Fatal error: Uncaught TypeError: implode(): Argument #2 ($array) must be of type ?array, string given in /customers/e/9/3/[website]/httpd.www/fotoalbum/plugins/Diagnostics/index.php:139
Stack trace:
#0 /customers/e/9/3/[website]/httpd.www/fotoalbum/plugins/Diagnostics/index.php(139): implode(Array, ', ')
#1 {main}
  thrown in /customers/e/9/3/[website]/httpd.www/fotoalbum/plugins/Diagnostics/index.php on line 139

And Log from Installation to first testalbum and log ins /out
2022-07-03 07:45:46 - error - Lychee\Modules\Database::createTables (180) 	- Table 'mumelter_netfoto.lychee_settings' doesn't exist
2022-07-03 07:45:46 - error - Lychee\Modules\Database::createTables (225) 	- Table 'mumelter_netfoto.lychee_albums' doesn't exist
2022-07-03 07:45:46 - error - Lychee\Modules\Database::createTables (248) 	- Table 'mumelter_netfoto.lychee_photos' doesn't exist
2022-07-03 07:49:08 - notice - Lychee\Modules\Photo::add (219) 	- Skipped adjustment of photo (01)
2022-07-03 07:49:08 - notice - Lychee\Modules\Photo::add (219) 	- Skipped adjustment of photo (03)
2022-07-03 07:49:09 - notice - Lychee\Modules\Photo::add (219) 	- Skipped adjustment of photo (05)
2022-07-03 07:49:09 - notice - Lychee\Modules\Photo::add (219) 	- Skipped adjustment of photo (04)
2022-07-03 07:49:10 - notice - Lychee\Modules\Photo::add (219) 	- Skipped adjustment of photo (06)
2022-07-03 07:49:11 - notice - Lychee\Modules\Photo::add (219) 	- Skipped adjustment of photo (DSC03385)
2022-07-03 07:49:12 - notice - Lychee\Modules\Photo::add (219) 	- Skipped adjustment of photo (07)
2022-07-03 07:49:13 - notice - Lychee\Modules\Photo::add (219) 	- Skipped adjustment of photo (09)
2022-07-03 07:49:13 - notice - Lychee\Modules\Photo::add (219) 	- Skipped adjustment of photo (11)
2022-07-03 07:49:13 - notice - Lychee\Modules\Photo::add (219) 	- Skipped adjustment of photo (DSC03390)
2022-07-03 07:49:14 - notice - Lychee\Modules\Photo::add (219) 	- Skipped adjustment of photo (DSC03389)
2022-07-03 07:49:15 - notice - Lychee\Modules\Photo::add (219) 	- Skipped adjustment of photo (DSC03391)
2022-07-03 07:49:15 - notice - Lychee\Modules\Photo::add (219) 	- Skipped adjustment of photo (DSC03395)
2022-07-03 07:49:16 - notice - Lychee\Modules\Photo::add (219) 	- Skipped adjustment of photo (DSC03397)
2022-07-03 07:49:17 - notice - Lychee\Modules\Photo::add (219) 	- Skipped adjustment of photo (DSC03398)
2022-07-03 07:49:17 - notice - Lychee\Modules\Photo::add (219) 	- Skipped adjustment of photo (DSC03399)
2022-07-03 07:49:18 - notice - Lychee\Modules\Photo::add (219) 	- Skipped adjustment of photo (DSC03401)
2022-07-03 07:49:20 - notice - Lychee\Modules\Photo::add (219) 	- Skipped adjustment of photo (DSC03404)
2022-07-03 07:49:21 - notice - Lychee\Modules\Photo::add (219) 	- Skipped adjustment of photo (DSC03405)
2022-07-03 07:49:21 - notice - Lychee\Modules\Photo::add (219) 	- Skipped adjustment of photo (DSC03406)
2022-07-03 07:59:38 - notice - Lychee\Modules\Session::login (81) 	- User (mumelter) has logged in from 31.207.99.238
2022-07-03 08:01:31 - notice - Lychee\Modules\Session::login (81) 	- User (mumelter) has logged in from 31.207.99.238
2022-07-03 08:19:17 - notice - Lychee\Modules\Session::login (81) 	- User (mumelter) has logged in from 31.207.99.238
2022-07-03 08:26:31 - notice - Lychee\Modules\Session::login (81) 	- User (mumelter) has logged in from 31.207.99.238

### Browser and system
Firefox and Mozilla on win11.

### PR DESCRIPTION
**Pull request by ildyria**
fixes #1341, fixes #1396

Add check if exec is available before executing it.
Looking forward to safe functions 2.0 which adds exec to the list.

### CODE REVIEW

File: database/migrations/2020_01_04_1200_config_has_ffmpeg.php [hunk 1]
```diff
@@ -18,6 +19,8 @@ public function up()
 		defined('BOOL') or define('BOOL', '0|1');
 		defined('TERNARY') or define('TERNARY', '0|1|2');
 
+		if (Helpers::isExecAvailable()) {
+			// Let's run the check for exiftool right here
```
Comment (d7415) [inline database/migrations/2020_01_04_1200_config_has_ffmpeg.php]: This looks like it was copied in error.

File: database/migrations/2020_01_04_1200_config_has_ffmpeg.php [hunk 2]
```diff
@@ -31,6 +34,9 @@ public function up()
 			$has_ffmpeg = 0;
 			// let's do nothing
 		}
```
Comment (d7415) [inline database/migrations/2020_01_04_1200_config_has_ffmpeg.php]: I know I'm normally the first to complain about whitespace patches, but I think this block should be indented.

**Review by kamil4 — APPROVED** [review]
LGTM

File: app/Actions/Diagnostics/Checks/ImageOptCheck.php [hunk 3]
```diff
@@ -36,9 +37,11 @@ public function check(array &$errors): void
...
 		foreach ($tools as $tool) {
-			$path = exec('command -v ' . $binaryPath . $tool->binaryName());
-			if ($path === '') {
-				$errors[] = 'Warning: lossless_optimization set to 1 but ' . $binaryPath . $tool->binaryName() . ' not found!';
+			if (Helpers::isExecAvailable()) {
+				$path = exec('command -v ' . $binaryPath . $tool->binaryName());
+				if ($path === '') {
+					$errors[] = 'Warning: lossless_optimization set to 1 but ' . $binaryPath . $tool->binaryName() . ' not found!';
+				}
 			}
```
Comment (nagmat84) [inline app/Actions/Diagnostics/Checks/ImageOptCheck.php]: I wonder if should actually check for a disabled `exec` in each iteration again and log one message for each iteration or if we should pull out the test for `exec`. In particular the way it is now, the user cannot distinguish between the case that the tool itself is missing and the case that the tool is yet there but `exec` is disabled. Users might get confused by this, if the error message (falsely) claims that the tool is missing.

File: app/Assets/Helpers.php [hunk 4]
```diff
@@ -257,4 +257,16 @@ public function data_index_set(int $idx = 0): void
...
+	 * Check if the `exec` function is available.
+	 *
+	 * @return bool
+	 */
+	public function isExecAvailable(): bool
+	{
+		$disabledFunctions = explode(',', ini_get('disable_functions'));
+
+		return function_exists('exec') && !in_array('exec', $disabledFunctions, true);
+	}
```
Comment (nagmat84) [inline app/Assets/Helpers.php:271]: Why do we need this function? Why can we not use the Laravel helper everywhere?

**Review by nagmat84 — COMMENTED** [review]
I left some questions.

File: app/Assets/Helpers.php (hunk 4 above)
Comment (ildyria) [reply app/Assets/Helpers.php:271]: What do you mean ?

Comment (nagmat84) [reply app/Assets/Helpers.php:271]: Sorry, my fault.

File: app/Actions/Diagnostics/Checks/ImageOptCheck.php (hunk 3 above)
Comment (d7415) [reply app/Actions/Diagnostics/Checks/ImageOptCheck.php]: > Users might get confused by this, if the error message (falsely) claims that the tool is missing.

But the error is only raised if exec is there? I haven't checked - does no `exec` raise a warning at the top of the page?

> or if we should pull out the test for `exec`.

That makes sense to me, putting the `if (Helpers::isExecAvailable()) {` outside the `foreach`.

Comment (nagmat84) [reply app/Actions/Diagnostics/Checks/ImageOptCheck.php]: Ah, sorry my fault. You are right, the error message is only logged _if_ `exec` is available. My fault. I reviewed this on my smartphone and the Github Android app is really bad at presenting diffs properly. I misread the curly braces.

Yes, the check for exec should be outside the loop.

**Pull request merged**